### PR TITLE
Partest tests can require a java version range

### PR DESCRIPTION
**Pull request by lrytz**
... using a `// javaVersion: N / N+ / N - M` comment in the test soruce.
The test is skipped if the java version is outside the range.

### CODE REVIEW

**Comment by som-snytt** [comment]
Ha, already looked at this when I woke up. Wondering if it means we can get rid of the "pragma" on the check file: control the version on the input side, not the output.

**Comment by lrytz** [comment]
I didn't add support for java version ranges to keep it simple, but that would also be a possibility. Getting rid of the pragmas in check files would then mean duplicating the test, which doesn't gain us much I think...

File: src/partest/scala/tools/partest/nest/Runner.scala [hunk 1]
```diff
@@ -533,17 +548,18 @@ class Runner(val testInfo: TestInfo, val suiteRunner: AbstractRunner) {
     // pass if it checks and didn't crash the compiler
     // or, OK, we'll let you crash the compiler with a FatalError if you supply a check file
     def checked(r: CompileRound) = r.result match {
+      case s: Skip => s
       case crash @ Crash(_, t, _) if !checkFile.canRead || !t.isInstanceOf[FatalError] => crash
-      case dnc @ _ => diffIsOk
+      case _ => diffIsOk
```
Comment (som-snytt) [inline src/partest/scala/tools/partest/nest/Runner.scala:564]: BTW this was intended to be "documentary". The underscore doesn't say why are we at the default case.

Also dnc is not Democratic National Committee but Does Not Compile.

File: src/partest/scala/tools/partest/nest/Runner.scala [hunk 2]
```diff
@@ -533,17 +548,18 @@ class Runner(val testInfo: TestInfo, val suiteRunner: AbstractRunner) {
...
     }
 
-    compilationRounds(testFile).find(!_.result.isOk).map(checked).getOrElse(genFail("expected compilation failure"))
+    compilationRounds(testFile).find(r => !r.result.isOk || r.result.isSkipped).map(checked).getOrElse(genFail("expected compilation failure"))
```
Comment (som-snytt) [inline src/partest/scala/tools/partest/nest/Runner.scala:567]: Always sad to lose an obscure use of underscore.

Previously, tests could not declare themselves skipped, but could only succeed trivially? There is "isJavaAtLeast" somewhere.

**Comment by som-snytt** [comment]
`--update-check` doesn't respect pragmas in check files, and probably never will.

There are only a few tests that are version or platform dependent. Maybe instead of range, just `jdk: 11 17` where someone testing locally with 16 will see a skipped test, which is reasonable.

**Comment by lrytz** [comment]
I agree that maintaining `.check` files with pragmas is a huge pain. Though switching jvm for a "quick" `--update-check` will be not so quick either :-)

I can do version ranges.

File: src/partest/scala/tools/partest/nest/Runner.scala [hunk 3]
```diff
@@ -514,9 +514,24 @@ class Runner(val testInfo: TestInfo, val suiteRunner: AbstractRunner) {
...
+  }
+
+  def compilationRounds(file: File): List[CompileRound] = {
+    import scala.tools.nsc.settings.ScalaVersion
+    import scala.util.Properties.javaSpecVersion
+    lazy val currentJavaVersion = ScalaVersion(javaSpecVersion)
+    val allFiles = sources(file)
+    val javaVersionOk = toolArgsFor(allFiles)("minJavaVersion").forall(v => {
+      val required = ScalaVersion(if (v.toInt <= 8) s"1.$v" else v)
+      currentJavaVersion >= required
```
Comment (som-snytt) [inline src/partest/scala/tools/partest/nest/Runner.scala]: `Properties.isJavaAtLeast(minJavaVersion)`, if that still works.

File: src/partest/scala/tools/partest/nest/Runner.scala [hunk 4]
```diff
@@ -514,9 +514,24 @@ class Runner(val testInfo: TestInfo, val suiteRunner: AbstractRunner) {
     def description = mkScalacString()
     lazy val result = { pushTranscript(description) ; attemptCompile(fs) }
   }
+  case class SkipRound(fs: List[File], reason: String) extends CompileRound {
+    def description: String = reason
+    lazy val result = Skip(testFile, reason)
```
Comment (som-snytt) [inline src/partest/scala/tools/partest/nest/Runner.scala]: aka `genSkip`

File: src/partest/scala/tools/partest/TestState.scala [hunk 5]
```diff
@@ -30,7 +30,7 @@ sealed abstract class TestState {
 
   def shortStatus    = if (isOk) "ok" else "!!"
 
-  final def andAlso(next: => TestState): TestState = if (isOk) next else this
+  final def andAlso(next: => TestState): TestState = if (isOk && !isSkipped) next else this
```
Comment (som-snytt) [inline src/partest/scala/tools/partest/TestState.scala:33]: `def hasNext = isOk && !isSkipped`? `isContinuing`?

**Review by som-snytt — APPROVED** [review]
`toolArgsForTheWin`

**Comment by lrytz** [comment]
Changed to support java version ranges, also added some doc to `CONTRIBUTING.md`

**Comment by harpocrates** [comment]
Thank you so much for adding this, and so promptly too. ❤️

**Pull request merged**